src/network/local_socket_reply.cpp

```diff
@@ -139,38 +143,105 @@ void mp::LocalSocketReply::send_request(const QNetworkRequest& request, QIODevic
-    http_data += "Host: multipass\r\n";
+    auto host = request.url().host().toUtf8();
+    if (!host.isEmpty())
+    {
+        http_data += "Host: " + host + "\r\n";
+    }
+    else
+    {
+        throw mp::HttpLocalSocketException("Host is required in URL");
+    }
 
     // Build the HTTP User-Agent header
-    http_data += "User-Agent: Multipass/";
-    http_data += mp::version_string;
-    http_data += "\r\n";
+    auto user_agent = request.header(QNetworkRequest::UserAgentHeader).toByteArray();
+    if (!user_agent.isEmpty())
+    {
+        http_data += "User-Agent: " + user_agent + "\r\n";
+    }
+
+    if (!local_socket_write(http_data))
+        return;
 
     if (op == "POST" || op == "PUT")
     {
-        http_data += "Content-Type: application/x-www-form-urlencoded\r\n";
+        http_data = "Content-Type: " + request.header(QNetworkRequest::ContentTypeHeader).toByteArray() + "\r\n";
 
-        if (outgoingData)
+        if (outgoingData && outgoingData->size() > 0)
         {
+            auto content_length = request.header(QNetworkRequest::ContentLengthHeader).toByteArray();
+            auto transfer_encoding = request.rawHeader("Transfer-Encoding");
+            bool is_chunked{false};
+
+            if (!content_length.isEmpty() && !transfer_encoding.isEmpty())
+            {
+                throw mp::HttpLocalSocketException(
+                    "Both \'Content-Length\' and \'Transfer-Encoding\' headers cannot be set at the same time");
+            }
+            else if (content_length.isEmpty() && transfer_encoding.isEmpty())
+            {
+                throw mp::HttpLocalSocketException(
+                    "Either the \'Content-Length\' or the \'Transfer-Encoding\' header must be set");
+            }
+            else if (!content_length.isEmpty())
+            {
+                http_data += "Content-Length: " + content_length + "\r\n";
+            }
+            else
+            {
+                http_data += "Transfer-Encoding: " + transfer_encoding + "\r\n";
+                is_chunked = true;
+            }
+
+            if (!local_socket_write(http_data + "\r\n"))
+                return;
+
+            local_socket->flush();
+
             outgoingData->open(QIODevice::ReadOnly);
+            std::vector<char> data_buffer;
+            data_buffer.reserve(max_bytes);
 
-            http_data += "Content-Length: ";
-            http_data += QByteArray::number(outgoingData->size());
-            http_data += "\r\n\r\n";
-            http_data += outgoingData->readAll();
+            while (true)
+            {
+                auto bytes_read = outgoingData->read(data_buffer.data(), max_bytes);
+                if (bytes_read < 0)
+                {
+                    throw mp::HttpLocalSocketException(
+                        fmt::format("Cannot read data to send to socket: {}", outgoingData->errorString()));
+                }
+
+                if (bytes_read == 0)
+                    break;
```

Comment:
I'd probably make it a:

```cpp
int bytes_read;
while (bytes_read = outgoingData->read(...) >= 0)
{
    ...
}

if (bytes_read < 0)
{
    throw ...
}
```

I suppose I'm allergic to `while(true)`, will leave it up to you :)

Reply:
It all accomplishes the same thing and I suppose is a coder's preference.  I'll think about it some and look at it from a readability perspective.

Reply:
Ok, I "caved" and did it your way :laughing: But it does make it more readable.